### ISSUE TITLE
Issues with io.on() typings

### ISSUE THREAD

**Issue by westy92** [socketio/socket.io#3833]
**Describe the bug**
After upgrading to socket.io 4.0.0, my code no longer compiles when using string enums for event names.

**To Reproduce**

Please fill the following code example:

Socket.IO server version: `4.0.0`
TypeScript@4.2.3

*Server*

```js
import { Server } from "socket.io";

const io = new Server(3000, {});

// this works fine
io.on("connection", (socket) => {});

enum Event {
    CONNECTION = "connection",
}
// Fails to compile!
// Argument of type '(socket:Socket) => void' is not assignable to parameter of type 'never'.
// ts(2345)
io.on(Event.CONNECTION, (socket) => {});
```

*Client*
N/A

**Expected behavior**
I expect my code to compile like it does for version 2.x and 3.x.

**Platform:**
N/A

**Comment by MaximeKjaer** [comment]
I'm able to replicate this, and I think I've found the root cause of the error. We use TypeScript conditional types internally to determine whether to get the listener type from a reserved events map, or from a user-provided events map. However, it seems that conditional types have some very weird behavior around enums. I've been able to minimize the issue to the following:

```ts
enum Events {
  CONNECTION = "connection",
  TEST = "test",
}
type Names = "test" | "connection";
type NameOrNull<N> = N extends Names ? N : null;

type test1 = NameOrNull<"test">; // returns type `"test"`
type test2 = NameOrNull<"not-a-name">; // returns type `null`
type test3 = NameOrNull<Events.TEST>; // returns type `never` (what???)
```

I would expect `test3` to be of type `Events.TEST`, but it comes out as type `never`, which is quite surprising. This could be a bug in the compiler. I'll try to do a bit more research into this, and if nothing comes up, I'll open an issue on the compiler.

Until we find a solution to this, there are two possible ways of dealing with this problem:

- Cast the enum member as `Event.CONNECTION as "connection"` to get TypeScript to treat it as a string literal type instead of an enum
- Migrate to strictly typed events by refactoring the `enum` to an `interface` that maps event names to event listener types, and passing that to the `Server` constructor

I realize that these are unfortunately less than ideal solutions :frowning_face:
